src/efibootdump.8
.TH "EFIBOOTDUMP" "8" "24 February 2016" "" ""

.SH NAME
efibootdump \- dump a boot entries from a variable or a file
.SH SYNOPSIS

\fBefibootdump\fR [\fB-?\fR|\fB--help\fR] [\fB--usage\fR]
.br
	[\fB-f\fR \fI<file1>\fR [... \fB-f\fR \fI<fileN>\fR]]
.br
	[[\fB-g\fR \fI{guid}\fR] \fI<name0>\fR [... [\fI<nameN>\fR]]]
.SH "DESCRIPTION"
.PP
\fBefibootdump\fR is a userspace application used to display individual UEFI boot options, from a file or a UEFI variable.  This allows e.g. saved files from efivarfs to be displayed, as well as variables on the running machine.

.SH "OPTIONS"
The following is a list of options accepted by efibootmgr:
.TP
\fB-g | --guid\fR \fI{guid}\fR
Any variables specified by name have the specified GUID.
.TP
\fB-f | --file\fR \fI<file>\fR
Read a single boot variable from the specified file.
.TP
\fI<nameN>\fR
Display the specified variable on the local machine.  If no GUID is specified, EFI Global Variable is the default.
.SH "BUGS"
.PP
Please direct any bugs, features, patches, etc. to the Red Hat installer team at https://github.com/rhboot/efibootmgr \&.
.SH "SEE ALSO"
.PP
efibootmgr(8)


doc/ChangeLog
commit de0ca446e7e5439306f7f54806d1a15310588426
Author: Jordan Hargrave <Jordan_Hargrave@dell.com>
Date:   Thu Nov 29 10:11:43 2012 -0600

    Add patches from fedora release
    * Wed Nov 28 2012 Matthew Garrett <mjg59@srcf.ucam.org> - 0.5.4-14
    - efibootmgr-0.5.4-Work-around-broken-Apple-firmware.patch
      Resolves: #873629
    - efibootmgr-0.5.4-Remove-device-path-padding-on-non-Itanium.patch - improve
      spec conformance

commit 9392250bd8668ef7544530d17800dc0271cd8bf3
Author: Jordan Hargrave <Jordan_Hargrave@dell.com>
Date:   Tue Apr 24 12:58:57 2012 -0500

    From: Lane Winner<lane.winner@oracle.com>
    
    make_boot_var does not check for failed status with create_variable.
    This can result in a memory leak.
    Additionally the user should be notified of the problem.
    
    We encounter this issue on one system after filling up the UEFI boot list
    with dummy devices.
    
    The patch fix the problem. It was verified on a Mensa system using RHEL 6.0
    
    Signed-off-by: Yinghai Lu<yinghai@kernel.org>

commit 8602b3c41701572669b1f8b9c45409d98405eca2
Author: Peter Jones <pjones@redhat.com>
Date:   Wed Jul 14 11:26:49 2010 -0700

    Handle sector_size != 512.
    
    Disks can have 4kB sectors now, so don't just bail out when that's the
    case.

commit fb3d9a8d9bfd580d23b14b384a393510e0a749ac
Author: Matt Domsch <Matt_Domsch@dell.com>
Date:   Thu Jul 23 14:20:19 2009 -0500

    fix disk minor number discovery
    
    Raymund Will noted disk_info_from_fd() incorrectly used logical &&
    instead of bitwise & when obtaining the minor number.
    
    Reported in https://bugzilla.novell.com/show_bug.cgi?id=524529#c1

commit acac7412e8e0819fc66cdadc06ad4ca535f29b35
Author: Matt Domsch <Matt_Domsch@dell.com>
Date:   Thu Jul 23 14:18:11 2009 -0500

    fix minor memory leak
    
    David Binderman noted new_data was being allocated but not freed.  Not
    a big deal as the program exits soon thereafter (and is thus freed),
    but worth fixing anyhow.
    
    Fixes https://bugzilla.novell.com/show_bug.cgi?id=524529#c1

commit c8d6ceaa7b33f952bcf32bc017ce8b5d7d659267
Author: Matt Domsch <Matt_Domsch@dell.com>
Date:   Fri Jan 11 15:09:12 2008 -0600

    update ChangeLog

commit 6e6bf6fc7665851798a6c2c92893ebb629e42aff
Author: Matt Domsch <Matt_Domsch@dell.com>
Date:   Fri Jan 11 15:08:12 2008 -0600

    replacing elilo < 3.6-6, not -5
    
    per conversation with clumens.

commit 4c1fd35da4d0478074e08c10225eb590576acf91
Author: Matt Domsch <Matt_Domsch@dell.com>
Date:   Thu Jan 3 12:19:28 2008 -0600

    RPM spec rewrite for Fedora inclusion
    
    efibootmgr is being split out of the elilo SRPM in Fedora, into its
    own SRPM.

commit 833cf4c1266ef72357948299008a22bfb80aa3f3
Author: Matt Domsch <Matt_Domsch@dell.com>
Date:   Thu Jan 3 12:18:31 2008 -0600

    Makefile cleanups
    
    bump version
    
    take an EXTRA_CFLAGS argument so rpmbuild can give us it's CFLAGS
    without overriding ours.
    
    exclude .git and *~ files from the tarball.

commit f0a8b91ba45ff4cf251805cc29aed4f8672c1801
Author: Matt Domsch <Matt_Domsch@dell.com>
Date:   Thu Jan 3 12:16:35 2008 -0600

    src/lib/efi.c: include linux/types.h
    
    patch from Fedora.

commit f387d5c5bde5d7129e41638e92faa2d38b7ad5a1
Author: Matt Domsch <Matt_Domsch@dell.com>
Date:   Thu Jan 3 12:15:18 2008 -0600

    make sure GPT_HEADER_SIGNATURE is a 64-bit value
    
    patch from Debian / Ubuntu 0.5.3-3ubuntu1.

commit 7b53efa45112f28e97451bdc16e6c6a68740bd79
Author: Matt Domsch <Matt_Domsch@dell.com>
Date:   Mon Nov 12 13:31:32 2007 -0600

    avoid more unaligned access warnings

commit 048197821f9ae2ef9e0c2bd4065649e72332e2dc
Author: Matt Domsch <Matt_Domsch@dell.com>
Date:   Mon Nov 12 12:25:42 2007 -0600

    cleanup a few unaligned access warnings

commit fa3942b34f1533659af8fe3f6fffb3b4acf4ee10
Author: Matt Domsch <Matt_Domsch@dell.com>
Date:   Mon Nov 12 12:12:37 2007 -0600

    cleanup exit values a little

commit c7e236783a79b6977df0ba03db0f92fabffc4b31
Author: Doug Chapman <doug.chapman@hp.com>
Date:   Mon Nov 12 11:32:12 2007 -0500

    patch to make efibootmgr have non-zero exit code on errors
    
    We have some automated tools that use efibootmgr which are having a hard
    time detecting when efibootmgr fails since it exits with 0 on many
    failures.  This patch catches (most) errors and exits with non-zero.  I
    also added several error messages for to make it more obvious what was
    wrong with the command line arguments.
    
    Signed-off-by: Matt Domsch <Matt_Domsch@dell.com>

commit ecd3c24cb6bee5072ff6d1292456ee3b2cc91019
Author: Matt Domsch <Matt_Domsch@dell.com>
Date:   Mon Jul 9 16:47:50 2007 +0000

    add -lz to libs, needed when libpci happens to need it

commit e192a055e0803263b71f89db732de73d5cf4de9b
Author: Matt Domsch <Matt_Domsch@dell.com>
Date:   Mon Jul 9 16:23:34 2007 +0000

    apply patch from Dave Jiang <djiang@mvista.com> to enable cross-building

commit 0ee8ecc10109b91d0a77098d5596f56780c862d8
Author: Matt Domsch <Matt_Domsch@dell.com>
Date:   Thu Aug 11 17:37:04 2005 +0000

    v0.5.2.2

commit 8097e44fbc1d022241837f318f57131b1df80adf
Author: Matt Domsch <Matt_Domsch@dell.com>
Date:   Thu Aug 11 17:36:29 2005 +0000

    document -@ option

commit c87e07e8e479220e696438dbeaf18ce73dafe2c8
Author: Matt Domsch <Matt_Domsch@dell.com>
Date:   Thu Aug 11 16:55:20 2005 +0000

    make efi_variable_t.DataSize be unsigned long, to match kernel implementation.  This isn't ideal, I'd prefer to be able to have a 32-bit efibootmgr on a 64-bit kernel, but with efi_status_t being a long also, we're already stuck with needing a native copy.

commit 5816bbbf56b37c9eee02a3539e99bffa8aec28e1
Author: Matt Domsch <Matt_Domsch@dell.com>
Date:   Wed Aug 10 15:59:33 2005 +0000

    clean up extra_opts_file handling

commit 8b02b12e1153f5bbf5010c16a5614464b5534c88
Author: Matt Domsch <Matt_Domsch@dell.com>
Date:   Wed Aug 10 15:28:30 2005 +0000


src/efibootdump.c
/*
 * efibootdump.c - dump a variable as if it's a Boot#### variable
 *
 * Copyright 2015 Red Hat, Inc.
 *
 * See "COPYING" for license terms.
 *
 * Author: Peter Jones <pjones@redhat.com>
 */

#include <ctype.h>
#include <efiboot.h>
#include <efivar.h>
#include <err.h>
#include <inttypes.h>
#include <libintl.h>
#include <locale.h>
#include <popt.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <uchar.h>
#include <unistd.h>

#include "error.h"
#include "unparse_path.h"

int verbose;

#define  _(String) gettext (String)
#define Q_(String) dgettext (NULL, String)
#define C_(Context,String) dgettext (Context,String)

static void
print_boot_entry(efi_load_option *loadopt, size_t data_size)
{
	char *text_path = NULL;
	size_t text_path_len = 0;
	uint8_t *optional_data = NULL;
	size_t optional_data_len = 0;
	uint16_t pathlen;
	const unsigned char *desc;
	char *raw;
	size_t raw_len;

	ssize_t rc;
	efidp dp = NULL;

	printf("%c ", (efi_loadopt_attrs(loadopt) & LOAD_OPTION_ACTIVE)
	              ? '*' : ' ');

	desc = efi_loadopt_desc(loadopt, data_size);
	if (!desc)
		printf("<invalid description> ");
	else if (desc[0])
		printf("%s ", desc);

	dp = efi_loadopt_path(loadopt, data_size);
	pathlen = efi_loadopt_pathlen(loadopt, data_size);

	rc = efidp_format_device_path(NULL, 0, dp, pathlen);
	if (rc < 0) {
		printf("<bad device path>");
		return;
	}
	text_path_len = rc + 1;
	text_path = alloca(text_path_len);
	if (!text_path)
		error(100, "Couldn't allocate memory");
	rc = efidp_format_device_path(text_path, text_path_len,
				      dp, pathlen);
	if (rc < 0) {
		printf("<bad device path>");
		return;
	}
	if (text_path && text_path_len >= 1)
		printf("%s", text_path);

	rc = efi_loadopt_optional_data(loadopt, data_size,
				       &optional_data, &optional_data_len);
	if (rc < 0) {
		printf("<bad optional_data>");
		return;
	}

	rc = unparse_raw_text(NULL, 0, optional_data, optional_data_len);
	if (rc < 0) {
		printf("<bad optional data>");
		return;
	}

	raw_len = rc + 1;
	raw = alloca(raw_len);
	if (!raw)
		error(101, "Couldn't allocate memory");

	rc = unparse_raw_text(raw, raw_len, optional_data, optional_data_len);
	if (rc < 0) {
		printf("<bad optional data>");
	} else if (rc > 0) {
		for (unsigned int i = 0; i < optional_data_len; i++)
			putchar(isprint(optional_data[i])
				? optional_data[i]
				: '.');
	}

	printf("\n");
}

int
main(int argc, char *argv[])
{
	const char **names = NULL;
	const char **files = NULL;
	char *guidstr = NULL;
	efi_guid_t guid = efi_guid_global;

	setlocale(LC_ALL, "");
	bindtextdomain("efibootdump", LOCALEDIR);
	textdomain("efibootdump");

	struct poptOption options[] = {
		{.argInfo = POPT_ARG_INTL_DOMAIN,
		 .arg = "efibootdump" },
		{.longName = "guid",
		 .shortName = 'g',
		 .argInfo = POPT_ARG_STRING |
			    POPT_ARGFLAG_OPTIONAL |
			    POPT_ARGFLAG_STRIP,
		 .arg = &guidstr,
		 .descrip = _("GUID namespace the variable is in"),
		 .argDescrip = "{guid}"},
		{.longName = "file",
		 .shortName = 'f',
		 .argInfo = POPT_ARG_ARGV |
			    POPT_ARGFLAG_OPTIONAL |
			    POPT_ARGFLAG_STRIP,
		 .arg = &files,
		 .descrip = _("File to read variable data from"),
		 .argDescrip = "<file>"},
		{.longName = "verbose",
		 .shortName = 'v',
		 .argInfo = POPT_ARG_VAL |
			    POPT_ARGFLAG_OPTIONAL |
			    POPT_ARGFLAG_STRIP,
		 .arg = &verbose,
		 .val = 2,
		 .descrip = _("Be more verbose on errors"),
		},
		POPT_AUTOALIAS
		POPT_AUTOHELP
		POPT_TABLEEND
	};
	efi_load_option *loadopt;
	uint8_t *data = NULL;
	size_t data_size = 0;

	poptContext optcon;
	optcon = poptGetContext("efibootdump", argc, (const char **)argv,
				options, 0);

	poptSetOtherOptionHelp(optcon, "[OPTIONS...] [name0 [... [nameN]]]");

	int rc;
	rc = poptReadDefaultConfig(optcon, 0);
	if (rc < 0 && !(rc == POPT_ERROR_ERRNO && errno == ENOENT))
		errorx(1, _("poptReadDefaultConfig failed: %s: %s"),
		       poptBadOption(optcon, 0), poptStrerror(rc));

	while ((rc = poptGetNextOpt(optcon)) > 0)
		;

	if (rc < -1)
		errorx(2, "Invalid argument: \"%s\": %s",
		       poptBadOption(optcon, 0), poptStrerror(rc));

	argc = poptStrippedArgv(optcon, argc, argv);
	names = poptGetArgs(optcon);
	if (!names && !files) {
		poptPrintUsage(optcon, stderr, 0);
		exit(4);
	}

	if (names && (!names[0] || names[0][0] == '\0')) {
		poptPrintUsage(optcon, stderr, 0);
		exit(4);
	}

	if (files && (!files[0] || files[0][0] == '\0')) {
		poptPrintUsage(optcon, stderr, 0);
		exit(4);
	}

	if (names) {
		if (guidstr) {
			rc = efi_id_guid_to_guid(guidstr, &guid);
			if (rc < 0)
				error(5, "Could not parse guid \"%s\"",
				      guidstr);
		}
		free(guidstr);
		guidstr = NULL;
		rc = efi_guid_to_str(&guid, &guidstr);
		if (rc < 0)
			error(6, "Guid lookup failed");
	}

	for (unsigned int i = 0;
	     files != NULL && files[i] != NULL && files[i][0] != '\0';
	     i++) {
		struct stat statbuf;
		FILE *f;
		size_t n;
		const char *filename = files[i];

		memset(&statbuf, 0, sizeof(statbuf));
		rc = stat(filename, &statbuf);
		if (rc < 0)
			error(7, "Could not stat \"%s\"", filename);

		data_size = statbuf.st_size;
		if (data_size == 0)
			errorx(11, "File \"%s\" is empty", filename);

		data = alloca(data_size);
		if (data == NULL)
			error(8, "Could not allocate memory");

		f = fopen(filename, "r");
		if (!f)
			error(9, "Could not open \"%s\"", filename);

		n = fread(data, 1, data_size, f);
		if (n < data_size)
			error(10, "Could not read \"%s\"", filename);

		printf("%s: ", filename);
		loadopt = (efi_load_option *)(data + 4);
		if (data_size <= 8)
			errorx(11, "Data is not a valid load option");
		if (efi_loadopt_is_valid(loadopt, data_size - 4)) {
			print_boot_entry(loadopt, data_size - 4);
		} else {
			loadopt = (efi_load_option *)data;
			if (!efi_loadopt_is_valid(loadopt, data_size))
				errorx(11, "Data is not a valid load option");
			print_boot_entry(loadopt, data_size);
		}

		fclose(f);
	}

	for (unsigned int i = 0;
	     names && names[i] != NULL && names[i][0] != '\0';
	     i++) {
		uint32_t attrs = 0;

		rc = efi_get_variable(guid, names[i], &data, &data_size,
				      &attrs);
		if (rc < 0) {
			warning("couldn't read variable %s-%s",
				names[i], guidstr);
			continue;
		}

		loadopt = (efi_load_option *)data;
		if (!efi_loadopt_is_valid(loadopt, data_size)) {
			warning("load option for %s is not valid", names[i]);
			printf("%d\n", __LINE__);
			if (data && data_size > 0) {
				free(data);
				continue;
			}
		}

		printf("%s", names[i]);
		if (efi_guid_cmp(&efi_guid_global, &guid))
			printf("-%s", guidstr);
		printf(": ");
		print_boot_entry(loadopt, data_size);
		if (data && data_size > 0)
			free(data);
	}

	if (guidstr)
		free(guidstr);

	poptFreeContext(optcon);
	return 0;
}
